ompi/mca/hook/comm_method/hook_comm_method_fns.c

```diff
@@ -0,0 +1,800 @@
+    for (i=0; i<nleaderranks; ++i) {
+        method[i] = comm_method(leader_comm, i);
+
+// For looking at our own local host though, we don't really want "self"
+// unless there's only one rank and "self" is the best answer. So if
+// there's more than one rank on our host, we get our local-host's
+// communication method for a neighbor on this host.
+        if (i == myleaderrank) {
+            if (nlocalranks > 1) {
+                method[i] = comm_method(local_comm, 1);
+            }
+        }
+    }
+
+// Gather the strings and the methods at rank 0.
+// The gatherv of the strings takes a few steps since we have to get
+// the sizes first and allocate the receiving string.
+    {
+        int len, *lens, *disps;
+
+        len = strlen(hoststring) + 1;
+        if (myleaderrank == 0) {
+            lens = malloc(nleaderranks * sizeof(int));
+            disps = malloc(nleaderranks * sizeof(int));
+        } else {
+            lens = disps = NULL;
+        }
+        leader_comm->c_coll->coll_gather(
+            &len, 1, MPI_INT,
+            lens, 1, MPI_INT,
+            0, leader_comm, leader_comm->c_coll->coll_gather_module);
+        if (myleaderrank == 0) {
+            int tlen = 0;
+            char *p;
+            for (i=0; i<nleaderranks; ++i) {
+                disps[i] = tlen;
+                tlen += lens[i];
+            }
+            allhoststrings = malloc(nleaderranks * sizeof(char*)  +  tlen);
+            p = (char*) (allhoststrings + nleaderranks);
+            for (i=0; i<nleaderranks; ++i) {
+                allhoststrings[i] = p;
+                p += lens[i];
+            }
+            leader_comm->c_coll->coll_gatherv(
+                hoststring, strlen(hoststring) + 1, MPI_CHAR,
+                &allhoststrings[0][0], lens, disps, MPI_CHAR,
+                0, leader_comm, leader_comm->c_coll->coll_gatherv_module);
+        } else {
+            // matching above call from rank 0, just &allhoststrings[0][0]
+            // isn't legal here, and those args aren't used at non-root anyway
+            leader_comm->c_coll->coll_gatherv(
+                hoststring, strlen(hoststring) + 1, MPI_CHAR,
+                NULL, NULL, NULL, MPI_CHAR,
+                0, leader_comm, leader_comm->c_coll->coll_gatherv_module);
+        }
+        if (myleaderrank == 0) {
+            free(lens);
+            free(disps);
+        }
+// and a simpler gather for the methods
+        leader_comm->c_coll->coll_gather(
+            method, nleaderranks, MPI_INT,
+            method, nleaderranks, MPI_INT,
+            0, leader_comm, leader_comm->c_coll->coll_gather_module);
+    }
+    ompi_comm_free(&local_comm);
+    ompi_comm_free(&leader_comm);
+
+// Interception for testing purposes. Let rank-0 meddle with all its method[]
+// settings, this is only for testing, eg to make sure the printing comes out
+// right.
+    if (myleaderrank == 0) {
+        p = getenv("MPI_COMM_METHOD_FAKEFILE");
```

Comment:
undocumented.

Reply:
Not only is this undocumented -- why wouldn't you use an MCA parameter for this?  (and the other place(s) where you use `getenv()`)